pkg/controller/operators/catalog/operator.go

```diff
@@ -1067,9 +1077,49 @@ func (o *Operator) ExecutePlan(plan *v1alpha1.InstallPlan) error {
 						if err != nil {
 							return errorwrap.Wrapf(err, "error find matched CSV: %s", step.Resource.Name)
 						}
+						crd.SetResourceVersion(currentCRD.GetResourceVersion())
 						if len(matchedCSV) == 1 {
-							// Attempt to update CRD
-							crd.SetResourceVersion(currentCRD.GetResourceVersion())
+							o.logger.Debugf("Found one owner for CRD %v", crd)
+							_, err = o.opClient.ApiextensionsV1beta1Interface().ApiextensionsV1beta1().CustomResourceDefinitions().Update(&crd)
+							if err != nil {
+								return errorwrap.Wrapf(err, "error updating CRD: %s", step.Resource.Name)
+							}
+						} else if len(matchedCSV) > 1 {
+							o.logger.Debugf("Found multiple owners for CRD %v", crd)
+
+							// ensure all old versions are present
+							for _, oldVersion := range currentCRD.Spec.Versions {
+								var versionPresent bool
+								for _, newVersion := range crd.Spec.Versions {
+									if oldVersion.Name == newVersion.Name {
+										versionPresent = true
+									}
+								}
+								if !versionPresent {
+									return fmt.Errorf("not allowing CRD (%v) update with unincluded version %v", crd.GetName(), oldVersion)
+								}
+							}
+
+							// no version bump
+							if len(currentCRD.Spec.Versions) == len(crd.Spec.Versions) {
+								for _, oldVersion := range currentCRD.Spec.Versions {
```

Comment:
since we have to do multiple passes over these versions, why not pull them into maps? e.g.  `oldVersions := map[string]json.Schema` (or whatever the schema type is)

Reply:
Your above suggestion led me to refactoring and removing this looping. I think for loops are clearer for what's happening, but see what you think.